tools/src/tensil/tools/compiler/Backend.scala

```diff
@@ -3,82 +3,195 @@
+    mutable.SortedMap.empty[BackendSegmentKey, BackendSegment]
 
-  def mkStream(
-      name: String,
-      stats: Option[BackendStats] = None
-  ): BackendStream =
-    new BackendStream(
-      name = name,
+  def mkSegment(
+      key: BackendSegmentKey,
+      stats: Option[Stats] = None
+  ): BackendSegment =
+    new BackendSegment(
+      key = key,
       layout = layout,
-      dataType = dataType,
-      printProgram = printProgramStream.isDefined,
-      printComments = printComments,
       stats = stats,
       tracepointConditions = tracepointConditions,
-      tracepointResolveRefToObject = tracepointResolveRefToObject
+      resolveRefToObject = resolveRefToObject
     )
 
-  def writeStream(
-      stream: BackendStream,
-      buffer: Array[Byte] = Array.fill[Byte](256)(0)
+  def emitSegment(segment: BackendSegment): Unit =
+    segments(segment.key) = segment
+
+  def writeSegments(
+      programStream: OutputStream,
+      printProgramFileName: Option[String] = None,
+      stats: Option[Stats] = None
   ): Unit = {
-    stream.closeAndWriteToOutputStream(programStream, buffer)
+    var instructionOffset: Long = 0
+    val writingLir = new lir.InstructionAddressInjector(
+      new lir.Broadcast(
+        Seq(
+          new lir.StreamGen(layout, programStream),
+          new lir.TracepointEmitter(traceContext)
+        ) ++ (if (printProgramFileName.isDefined)
+                Seq(
+                  new lir.Printer(
+                    printProgramFileName.get
+                  )
+                )
+              else
+                Nil) ++ (if (stats.isDefined)
+                           Seq(
+                             new lir.StatsGen(
+                               layout.arch,
+                               stats.get
+                             )
+                           )
+                         else
+                           Nil): _*
+      )
+    )
 
-    if (printProgramStream.isDefined) {
-      val tb = new TablePrinter(Some(stream.name))
+    val windowSize = layout.arch.numberOfThreads match {
+      case 1 => 1
+      case 2 => 3
+      case _ =>
+        throw new CompilerException(
+          s"${layout.arch.numberOfThreads} threads are not supported"
+        )
+    }
+
+    var nextTid = 0
 
-      var i = 0
-      for (instructionLine <- stream.instructionLines) {
-        val printProgramOffset = instructionOffset + i
+    case class ThreadedPartition(
+        init: Option[BackendSegment] = None,
+        load: Option[BackendSegment] = None,
+        compute: Option[BackendSegment] = None,
+        save: Option[BackendSegment] = None,
+    ) {
+      val tid = nextTid
+
+      nextTid = (nextTid + 1) % layout.arch.numberOfThreads
+    }
 
-        tb.addLine(
-          TableLine(
-            s"P[$printProgramOffset]",
-            instructionLine
+    var curInit: Option[BackendSegment] = None
+    val partitions = Seq.fill(windowSize - 1)(ThreadedPartition()) ++
```

Comment:
We use `var` to allocate TID for each `ThreadedPartition` by incrementing `nextTid` and instantiating `ThreadedPartition` in a strict order of layer/stage/partition from segment key. We also pad with `windowSize - 1` empty `ThreadedPartition` between layers to ensure inter-layer partitions are not parallelized because they have data dependency as well as that first and last window will process first and last segments correctly.

TID allocation could probably be expressed in more functional style.

Reply:
Probably the best place to properly explain this would be in the Compiler section of documentation. Few diagrams would help a lot.

Reply:
Added some comments in the code.

Reply:
Also got rid of `var nextTid`

Reply:
Added even more comments. Ptyl.